File: app/android/app/src/main/java/com/radiostream/player/SongFactory.java
package com.radiostream.player;

import android.content.Context;

import com.radiostream.Settings;
import com.radiostream.networking.metadata.MetadataBackendGetter;
import com.radiostream.networking.models.SongResult;

import javax.inject.Inject;

public class SongFactory {

    private MediaPlayerFactory mMediaPlayerFactory;
    private Context mContext;
    private Settings mSettings;
    private MetadataBackendGetter mMetadataBackendGetter;

    @Inject
    public SongFactory(MediaPlayerFactory mediaPlayerFactory, Context context, Settings settings,
                       MetadataBackendGetter metadataBackend) {

        mMediaPlayerFactory = mediaPlayerFactory;
        mContext = context;
        mSettings = settings;
        mMetadataBackendGetter = metadataBackend;
    }

    public Song build(SongResult songResult) {
        return new Song(songResult, mMediaPlayerFactory.build(), mContext, mSettings,
            mMetadataBackendGetter);
    }

    public Song build(Song otherSong) {
        return new Song(otherSong, mMediaPlayerFactory.build(), mContext, mSettings,
            mMetadataBackendGetter);
    }
}

File: app/android/app/src/main/java/com/radiostream/player/SongFactory.kt
package com.radiostream.player

import android.content.Context

import com.radiostream.Settings
import com.radiostream.networking.metadata.MetadataBackendGetter
import com.radiostream.networking.models.SongResult
import com.radiostream.wrapper.UriInterface
import com.radiostream.wrapper.UriWrapper

import javax.inject.Inject

class SongFactory @Inject
constructor(private val mMediaPlayerFactory: MediaPlayerFactory, private val mContext: Context, private val mSettings: Settings,
            private val mMetadataBackendGetter: MetadataBackendGetter, private val mUriWrapper: UriWrapper) {

    fun build(songResult: SongResult): Song {
        return Song(songResult, MediaPlayerWrapper(mMediaPlayerFactory.build()), mContext, mSettings,
                mMetadataBackendGetter, mUriWrapper)
    }

    fun build(otherSong: Song): Song {
        return Song(otherSong, MediaPlayerWrapper(mMediaPlayerFactory.build()), mContext, mSettings,
                mMetadataBackendGetter, mUriWrapper)
    }
}
